### PR TITLE
Member codes docs

### PR DESCRIPTION
**Pull request by elichad**
Addresses #2576 except for the admin guide updates.

Merge when ready to release v4.3.

### CODE REVIEW

File: docs/design/design_patterns.md [hunk 1]
```diff
@@ -39,6 +39,17 @@ This document acts as a reference for where certain design patterns can be found
...
+2. On `Form`s connected to the model, manipulate the `helper.layout` to show/hide the new field according to the value of the field you're implementing soft validation on (e.g. the `validate_member_code` method of `TrainingRequestForm` in [`/extforms/forms.py`](../../amy/extforms/forms.py) shows and hides the `member_code_override` field according to the validity of the `member_code`).
+3. Build validation carefully for the override and the field it relates to. The override should only be required and `True` if the related field is invalid. In other cases, it should be `False` - this may require updating the value during validation (e.g. the `validate_member_code` method again).
+4. Consider adding a filter to help admins find objects where the override was used. Beware that the default `django_filters.BooleanFilter` is not quite appropriate - typically you will want *all* results to be shown when the filter is `False`, and only results that use the override to be shown when the filter is `True` (e.g. `invalid_member_code` filter in `TrainingRequestFilter` in [/extrequests/filters.py](../../amy/extrequests/filters.py)).
```
Comment (pbanaszkiewicz) [inline docs/design/design_patterns.md:51]: Consider absolute link to the file in GitHub repository.

File: docs/design/design_patterns.md [hunk 2]
```diff
@@ -39,6 +39,17 @@ This document acts as a reference for where certain design patterns can be found
...
+
+**Demo**: On test-amy or a local AMY instance, go to the Instructor Training application form and select "Profile Creation for Pre-approved Trainees." Enter some random letters as the registration code (an invalid input), then fill out the rest of the required form fields. After clicking "Submit", if member code enforcement is [enabled](./projects/2023_member_code_enforcement.md#feature-flag), you should see an error on the registration code starting with "This code is invalid." Underneath the field, a checkbox should be visible with the label "Continue with registration code marked as invalid" - if you check this box, you should now be able to submit the form.
+
+**Method**:
+
+1. Add a field `<fieldname>-override` to the relevant model. This should be a `BooleanField` defaulting to `False`.
+2. On `Form`s connected to the model, manipulate the `helper.layout` to show/hide the new field according to the value of the field you're implementing soft validation on (e.g. the `validate_member_code` method of `TrainingRequestForm` in [`/extforms/forms.py`](../../amy/extforms/forms.py) shows and hides the `member_code_override` field according to the validity of the `member_code`).
```
Comment (pbanaszkiewicz) [inline docs/design/design_patterns.md:49]: Consider absolute link to the file in GitHub repository.

File: docs/design/projects/2023_member_code_enforcement.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+[GitHub project](https://github.com/carpentries/amy/projects/15).
+
+In early 2024, the membership team planned to start enforcing that all preapproved IT applications and workshop requests should include a valid and active member code (a.k.a. registration code, group code)
+
+In mid-2023, the checkout requirements for Instructor Training were updated. As the relevant changes were made to AMY to support this update, the project also expanded to include improvements to the training progress view for trainees.
+
+## Feature flag
+
+The `ENFORCE_MEMBER_CODES` [feature flag](../feature_flags.md) controls whether validity checks are performed on member codes. If set to `True` in the Django admin panel, validity checks will be performed on all submissions. If set to `False`, no checks will be performed (i.e. all codes will be accepted). The value of the flag does **not** affect whether the `member_code` field is displayed on forms or in views.
```
Comment (pbanaszkiewicz) [inline docs/design/projects/2023_member_code_enforcement.md:11]: Need to switch the feature flag `ENFORCE_MEMBER_CODES` on before release 4.3.

**Review by pbanaszkiewicz — APPROVED** [review]

**Comment by pbanaszkiewicz** [comment]
Approved. This should be merged before releasing 4.3. Some additional changes were mentioned during the review.

**Comment by pbanaszkiewicz** [comment]
@maneesha @froggleston are we releasing member codes in v4.3? If so, I suppose I should work on resolving conflicts here and eventually merging this PR?

**Comment by maneesha** [comment]
We have not had a chance to test any of this so I think we should wait on this.

**Comment by pbanaszkiewicz** [comment]
Merging in v4.3 since this contains instructions for the upcoming deployment and the project was merged and deployed to test environment.

File: docs/design/projects/2023_member_code_enforcement.md (hunk 3 above)
Comment (pbanaszkiewicz) [reply docs/design/projects/2023_member_code_enforcement.md:11]: Done: https://github.com/carpentries/amy/commit/506a2fe544436d48389ae2e01295dae34ec24b41

**Pull request merged**